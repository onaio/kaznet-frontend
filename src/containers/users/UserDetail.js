import React, { Component } from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";

import * as userSelectors from "../../store/users/reducer";
import * as userActions from "../../store/users/actions";
import * as globalActions from "../../store/global/actions";
import * as errorHandlerSelectors from "../../store/errorHandler/reducer";

import NestedElementMap from "../NestedElementMap";
import DetailView from "../../components/DetailView";
import UserDetailTitle from "../../components/users/UserDetailTitle";

export class UserDetail extends Component {
  componentDidMount() {
    this.props.fetchUser(this.props.match.params.id);
    this.props.noTitle();
  }

  render() {
    this.user = this.props.usersById;
    if (!this.user) return this.renderLoading();
    return (
      <div className="UsersList">
        <UserDetailTitle user={this.user} />
        <DetailView
          renderMainDetails={this.renderMainDetails()}
          renderAdditionalDetails={this.renderAdditionalDetails()}
        />
      </div>
    );
  }

  renderLoading() {
    if (!this.props.hasError) {
      return <p>Loading...</p>;
    } else if (this.props.hasError) {
      return <p> {this.props.errorMessage.message} </p>;
    }
  }

  renderMainDetails() {
    const headerItems = {
      Username: this.user.attributes.ona_username,
      Role: this.user.attributes.role_display,
      "National ID": this.user.attributes.national_id
    };

    return <NestedElementMap detailitems={headerItems} HTMLTag="td" />;
  }

  renderAdditionalDetails() {
    const headerItems = {
      Gender: this.user.attributes.gender,
      Address: this.user.attributes.address,
      "Level of Expertise": this.user.attributes.expertise,
      Email: this.user.attributes.email,
      "Phone Number": this.user.attributes.phone_number
    };

    return <NestedElementMap detailitems={headerItems} HTMLTag="td" />;
  }
}

function mapStateToProps(state, props) {
  return {
    usersById: userSelectors.getUserById(state, props.match.params.id),
    hasError: errorHandlerSelectors.getHasError(state),
    errorMessage: errorHandlerSelectors.getErrorMessage(state)
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      fetchUser: userActions.fetchUser,
      noTitle: globalActions.toggleTitleOff
    },
    dispatch
  );
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(UserDetail);
