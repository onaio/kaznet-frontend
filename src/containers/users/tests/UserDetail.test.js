// Test UserDetail
import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import { Router } from "react-router";
import createBrowserHistory from "history/createBrowserHistory";

import { UserDetail } from "../UserDetail";
import * as fixtures from "../../../store/users/tests/fixtures";
import ErrorBoundary from "../../../components/ErrorBoundary";

const history = createBrowserHistory();

describe("containers/users/UserDetail", () => {
  it("renders without crashing", () => {
    shallow(
      <UserDetail
        match={{
          params: {
            id: "1"
          }
        }}
        fetchUser={function() {}}
        userById={fixtures.userById}
        noTitle={function() {}}
      />
    );
  });

  it("renders detail page correctly without delete button", () => {
    const wrapper = mount(
      <Router history={history}>
        <ErrorBoundary>
          <UserDetail
            match={{
              params: {
                id: "1"
              }
            }}
            fetchUser={function() {}}
            noTitle={function() {}}
            userById={fixtures.userById}
          />
        </ErrorBoundary>
      </Router>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });

  it("renders detail page correctly with delete button", () => {
    const wrapper = mount(
      <Router history={history}>
        <ErrorBoundary>
          <UserDetail
            match={{
              params: {
                id: "7"
              }
            }}
            fetchUser={function() {}}
            noTitle={function() {}}
            userById={fixtures.userById}
          />
        </ErrorBoundary>
      </Router>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
