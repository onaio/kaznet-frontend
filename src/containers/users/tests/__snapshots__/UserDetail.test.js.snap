// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/users/UserDetail renders detail page correctly with delete button 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <ErrorBoundary>
    <UserDetail
      fetchUser={[Function]}
      match={
        Object {
          "params": Object {
            "id": "7",
          },
        }
      }
      noTitle={[Function]}
      userById={
        Object {
          "attributes": Object {
            "amount_earned": "0 KES",
            "approval_rate": 0,
            "approved_submissions": 0,
            "avg_amount_earned": 0,
            "avg_approval_rate": 0,
            "avg_approved_submissions": 0,
            "avg_rejected_submissions": 0,
            "avg_submissions": 0,
            "created": "2018-06-19T16:03:10.184043+03:00",
            "email": "sol@admin.me",
            "expertise": "1",
            "first_name": "Davis",
            "gender": "1",
            "last_login": "2018-06-19T16:03:22+03:00",
            "last_name": "Raymond",
            "modified": "2018-06-20T08:45:12.976916+03:00",
            "national_id": null,
            "ona_pk": null,
            "ona_username": "davisraym",
            "payment_number": "",
            "phone_number": "",
            "rejected_submissions": 0,
            "role": "1",
            "role_display": "Admin",
            "submission_count": 0,
          },
          "id": "1",
          "type": "UserProfile",
        }
      }
    >
      <p>
        Loading...
      </p>
    </UserDetail>
  </ErrorBoundary>
</Router>
`;

exports[`containers/users/UserDetail renders detail page correctly without delete button 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <ErrorBoundary>
    <UserDetail
      fetchUser={[Function]}
      match={
        Object {
          "params": Object {
            "id": "1",
          },
        }
      }
      noTitle={[Function]}
      userById={
        Object {
          "attributes": Object {
            "amount_earned": "0 KES",
            "approval_rate": 0,
            "approved_submissions": 0,
            "avg_amount_earned": 0,
            "avg_approval_rate": 0,
            "avg_approved_submissions": 0,
            "avg_rejected_submissions": 0,
            "avg_submissions": 0,
            "created": "2018-06-19T16:03:10.184043+03:00",
            "email": "sol@admin.me",
            "expertise": "1",
            "first_name": "Davis",
            "gender": "1",
            "last_login": "2018-06-19T16:03:22+03:00",
            "last_name": "Raymond",
            "modified": "2018-06-20T08:45:12.976916+03:00",
            "national_id": null,
            "ona_pk": null,
            "ona_username": "davisraym",
            "payment_number": "",
            "phone_number": "",
            "rejected_submissions": 0,
            "role": "1",
            "role_display": "Admin",
            "submission_count": 0,
          },
          "id": "1",
          "type": "UserProfile",
        }
      }
    >
      <p>
        Loading...
      </p>
    </UserDetail>
  </ErrorBoundary>
</Router>
`;
